### ISSUE TITLE
Bogus FL_MOVE/FL_DRAG events

### ISSUE THREAD

**Issue by CendioOssman** [fltk/fltk#76]
We've unfortunately found an issue where this code will result in bogus FL_MOVE/FL_DRAG events being sent:

https://github.com/fltk/fltk/blob/f386bd2cb78fc9a5bffdf0451870720aafacca37/src/Fl_x.cxx#L221-L224

What happens is that we get a `Leave` which sets `fl_xmousewin` to `NULL`. Any X11 event after that will cause the above code to be run as `send_motion` will also be `NULL`.

I suspect the issue is simply that an extra `NULL` check is needed as I assume this code was only meant to be run if there was an actual window to send things to.

**Comment by Albrecht-S** [comment]
Thanks for reporting this. Do you have a demo program that shows the issue?

FTR: `Fl_cocoa.mm` has similar code and has an additional `NULL` check:
https://github.com/fltk/fltk/blob/f386bd2cb78fc9a5bffdf0451870720aafacca37/src/Fl_cocoa.mm#L757-L761

although `CONSOLIDATE_MOTION` is currently set to 0 in `Fl_cocoa.mm` vs. 1 in `Fl_x.cxx`.

Anyway -- without deeper analysis -- it appears to be _plausible_ to add the same check in Fl_x.cxx, but I'd like to check this further...

**Comment by Albrecht-S** [comment]
@CendioOssman What happens if you add this NULL check? Does it fix the issue? Is this issue at all reproducible for you in any way?

**Comment by CendioOssman** [comment]
Our test case is a bit complex in that we're adding support for touch events as well. Unfortunately I don't have a different test case at the moment.

We did not try changing that code, but I can confirm via gdb that the value is indeed `NULL`.

(I'd also prefer for `CONSOLIDATE_MOTION` to be off by default. It doesn't seem to handle all corner cases properly so I think it should be opt-in.)

**Comment by Albrecht-S** [comment]
Thanks for your quick reply.

I agree that options like `CONSOLIDATE_MOTION` should probably be settable by the user. Changing defaults (making it opt-**in**) would be something that needs to be considered carefully - although it appears to be the opposite on macOS and I didn't find it at all for Windows code (if it applies to that as well).

A few more questions: can you tell what exactly `CONSOLIDATE_MOTION` does? What's the difference if it is turned off? The reason I'm asking is because I assume you could share some knowledge which I currently don't have and would need to find out by studying the code and/or testing. TIA for answers.

**Comment by Albrecht-S** [comment]
PS: what are the corner cases that are not handled properly? Can you elaborate?

**Comment by CendioOssman** [comment]
The purpose of `CONSOLIDATE_MOTION` seems to be to avoid bothering the application with lots and lots of `FL_MOVE` events. So if several X11 `MotionNotify` are received then it only reports the last of them. So there is an assumption here that a) we'll get many events from the X11 server, and b) that the application will have problems dealing with them.

One obvious corner case is what happens if you get `MotionNotify` for two different windows? The current code can only track one window so any other events are simply lost.

FLTK's event model is also not really designed for sending events later. What if you get a mix of `MotionNotify` and other events? Will the state fields and other things be correct once that delayed `FL_MOVE` is sent?

**Comment by Albrecht-S** [comment]
Note to myself:
This issue requires `CMake` (and optionally autoconf/configure) support to make `CONSOLIDATE_MOTION` user selectable. Another way to achieve configurability would be to define the `CONSOLIDATE_MOTION` macro only if it has not been defined by the user (on the compiler command line), something like:
```
#ifndef CONSOLIDATE_MOTION
  #define CONSOLIDATE_MOTION (0)
#endif
```
This would also set the default `OFF` (0) for consistency and let the user compile with
`-DCONSOLIDATE_MOTION=1` if desired.

This needs to be documented, together with the exact implications...

Or we could remove this option entirely...

Comments welcome.

**Comment by Albrecht-S** [comment]
The two commits some minutes ago are in Git `master` and `branch-1.3`, respectively.

This fixes the issue reported here but does not change anything WRT adding an option to disable CONSOLIDATE_MOTION w/o editing the sources (or to disable it (default) with opt-in). This will be addressed later (in 1.4 only) and this is the reason why I'm leaving this issue open but added the label "fixed".

@CendioOssman: Thanks for the report and the proposed fix.

Note to myself: **this issue** should be closed and another issue addressing the compiler/preprocessor/build option should probably be opened...

**Comment by Albrecht-S** [comment]
@CendioOssman Is this issue essential for TigerVNC?

**Comment by CendioOssman** [comment]
No, it's not. We added a workaround in TigerVNC/tigervnc@6e8147eb122eb8670513d6001e413bd983e93075. It's appreciated a fix is in though so we can remove the workaround eventually.

**Comment by Albrecht-S** [comment]
OK, I committed final fixes to both branches (1.3 and 1.4 (master)).

Summary:
- The original issue (NULL check) had already been fixed in both branches on Apr 26, 2021 and has been included in FLTK 1.3.6.
- My recent commits renamed the macro `CONSOLIDATE_MOTION` to `FLTK_CONSOLIDATE_MOTION` and made it user-settable via compiler commandline as suggested above.

The default value in 1.3 is still 1 (ON) for backwards compatibility whereas it is now 0 (OFF) in FLTK 1.4.0. I suggest not to change it though unless you really need it and you know what you're doing (the resulting code is not tested as well as the default setting).

I'm closing this now since it's entirely fixed.
